### PR TITLE
Don't load the media repo when configured to use an external media repo

### LINKED ISSUE #5517: Turning off the media repo doesn't stop the background updates from running
update_recently_accessed_media and expire_url_cache_data don't need to run.

### PR DESCRIPTION
**Pull request by hawkowl**
This checks if directories exist, which does not matter when the media repo is used.

Fixes #5517

### CODE REVIEW

File: synapse/config/repository.py [hunk 1]
```diff
@@ -202,6 +215,21 @@ def generate_config_section(self, data_dir_path, **kwargs):
...
+        ##
+        ## Media Store
+        ##
+
+        # Is an external media repo in use?
+        #
+        # Set 'False' to use Synapse's media repo. If this is set to 'True',
+        # Synapse's media repo below must be disabled.
+        #
+        #external_media_repo: False
```
Comment (richvdh) [inline synapse/config/repository.py]: this should be ready for uncommenting:

```suggestion
        #external_media_repo: True
```

File: synapse/config/repository.py [hunk 2]
```diff
@@ -202,6 +215,21 @@ def generate_config_section(self, data_dir_path, **kwargs):
 
         return (
             r"""
```
Comment (richvdh) [inline synapse/config/repository.py]: can you remove the spurious newline while you are here:

```suggestion
            r"""\
```

File: synapse/config/repository.py [hunk 3]
```diff
@@ -202,6 +215,21 @@ def generate_config_section(self, data_dir_path, **kwargs):
...
+        ##
```
Comment (richvdh) [inline synapse/config/repository.py]: this is the wrong style, should be:

```
## Media store ##
```

File: synapse/config/repository.py [hunk 4]
```diff
@@ -202,6 +215,21 @@ def generate_config_section(self, data_dir_path, **kwargs):
...
+        ## Media Store
+        ##
+
+        # Is an external media repo in use?
```
Comment (richvdh) [inline synapse/config/repository.py]: It seems very confusing to have two settings, rather than a single `enable_media_repo`. Can we not combine them?

File: synapse/config/repository.py [hunk 5]
```diff
@@ -202,6 +215,21 @@ def generate_config_section(self, data_dir_path, **kwargs):
...
+        #
+        # Set 'False' to use Synapse's media repo. If this is set to 'True',
```
Comment (richvdh) [inline synapse/config/repository.py]: there's a bit of a mix, but we use true/false more than True/False.

File: synapse/config/repository.py [hunk 6]
```diff
@@ -202,6 +215,21 @@ def generate_config_section(self, data_dir_path, **kwargs):
...
+        # Synapse's media repo below must be disabled.
+        #
+        #external_media_repo: False
+
+        # Enable the media store service inside Synapse.
+        #
+        enable_media_repo: True
```
Comment (richvdh) [inline synapse/config/repository.py]: config settings should be commented out, as a rule, and given an example value which might be useful (ie, the opposite to the default).

File: synapse/config/repository.py (hunk 2 above)
Comment (richvdh) [inline synapse/config/repository.py]: All this stuff is docced at https://github.com/matrix-org/synapse/blob/develop/docs/code_style.rst#configuration-file-format btw

File: synapse/config/repository.py (hunk 3 above)
Comment (hawkowl) [reply synapse/config/repository.py]: fixed

File: docs/sample_config.yaml [hunk 7]
```diff
@@ -565,6 +565,12 @@ log_config: "CONFDIR/SERVERNAME.log.config"
 
 
 
+## Media Store ##
+
+# Enable the media store service inside Synapse.
+#
+# enable_media_repo: False
```
Comment (anoadragon453) [inline docs/sample_config.yaml]:
```suggestion
# enable_media_repo: false
```

pls

File: synapse/app/media_repository.py [hunk 8]
```diff
@@ -71,13 +73,20 @@ def _listen_http(self, listener_config):
...
+                    register_servlets_for_media_repo(self, admin_resource)
+
                     resources.update(
                         {
                             MEDIA_PREFIX: media_repo,
                             LEGACY_MEDIA_PREFIX: media_repo,
                             CONTENT_REPO_PREFIX: ContentRepoResource(
                                 self, self.config.uploads_path
                             ),
+                            "/_synapse/admin": admin_resource,
```
Comment (richvdh) [inline synapse/app/media_repository.py]: this seems to be sprawling a bit beyond the PR title and changelog comment. Please could the admin stuff be kept separate?

File: synapse/config/repository.py [hunk 9]
```diff
@@ -202,6 +215,12 @@ def generate_config_section(self, data_dir_path, **kwargs):
 
         return (
             r"""
+        ## Media Store ##
+
+        # Enable the media store service inside Synapse.
+        #
+        # enable_media_repo: False
```
Comment (richvdh) [inline synapse/config/repository.py]:
```suggestion
        #enable_media_repo: false
```

File: synapse/config/repository.py [hunk 10]
```diff
@@ -202,6 +215,12 @@ def generate_config_section(self, data_dir_path, **kwargs):
 
         return (
             r"""
+        ## Media Store ##
+
+        # Enable the media store service inside Synapse.
```
Comment (richvdh) [inline synapse/config/repository.py]:
```suggestion
        # Enable the media store service in the Synapse master. Uncomment the following if you are using a separate
        # media store worker.
```

File: synapse/rest/admin/__init__.py [hunk 11]
```diff
@@ -36,36 +36,23 @@
...
+    QuarantineMediaInRoom,
+)
 from synapse.rest.admin.server_notice_servlet import SendServerNoticeServlet
 from synapse.types import UserID, create_requester
 from synapse.util.versionstring import get_version_string
 
 logger = logging.getLogger(__name__)
 
 
-def historical_admin_path_patterns(path_regex):
```
Comment (richvdh) [inline synapse/rest/admin/__init__.py]: again, please can the admin stuff go in a separate PR?

File: synapse/app/media_repository.py (hunk 8 above)
Comment (richvdh) [reply synapse/app/media_repository.py]: ohhh I see, it's the admin APIs that pull in the `MediaRepository` which runs the background jobs. sorry, ignore this, I'll have another look.

File: synapse/rest/admin/__init__.py [hunk 12]
```diff
@@ -822,10 +746,18 @@ def register_servlets(hs, http_server):
     VersionServlet(hs).register(http_server)
 
 
+def register_servlets_for_media_repo(hs, http_server):
```
Comment (richvdh) [inline synapse/rest/admin/__init__.py]: suggestion: could move this into `admin/media.py`?
